### DISCUSSION TITLE
Configuration Files

### DISCUSSION THREAD

**Discussion by dmarro89** [dmarro89/dare-db#8] [Ideas]
What about the database configuration files?

**Comment by vdmitriyev** [comment]
### Configuration Concept

Here is an initial concept how a configuration file could work:

* Name of the configuration file should be "config.toml" 
* Configuration file should be located at the root folder of the database
* Configuration file should be generated on the initial start (if not such file exists)
* Configuration file should use ```TOML``` format
* Configuration values should be read as follows:
	+ ```<CATEGORY><CONFIGURATION>```
	+ Ex. "server.host"
* There is should be a possibility to override configuration values from the file using "environment variables". It works as follows:
    1. On start, a configuration file should be read and the configs parameters should be set
    2. Then environment variables should be red-read and re-write the original configs values supplied within the configuration file

### Configuration Parameters and Default Values

The database should use a configuration file that will defined following parameters:

| **Configuration** | **Default** | **ENV Name** | **Description** | **Category**| **Status**|
|---|---|---|---|---|---|
| host| 127.0.0.1 | DARE_HOST | Name or IP address of the server | server | NOT IMPLEMENTED|
| port | 2605 | DARE_PORT | Port number for the server | server |NOT IMPLEMENTED|
| admin_user | admin | DARE_USER | Username for the admin user (future implementation) | server | NOT IMPLEMENTED|
| admin_password | <**GENERATED-ON-START**> | DARE_PASSWORD | Password for the admin user (future implementation) | server |NOT IMPLEMENTED|
| log_level | INFO | DARE_LOG_LEVEL | Log level | log | NOT IMPLEMENTED|
| log_file | daredb.log | DARE_LOG_FILE | Path to the desired location for log files (optional) | log | NOT IMPLEMENTED|
| data_dir | data | DARE_DATA_DIR | Path to the desired location to persist data (if required) | settings | NOT IMPLEMENTED|
| tls_enabled | False | DARE_TLS_ENABLED | Enables HTTPS to access data | security | NOT IMPLEMENTED|
| tls_cert_private | <**ROOT**>/settings/cert_private.pem | DARE_TLS_CERT_PRIVATE | Private certificate for secure connection | security | NOT IMPLEMENTED|
| tls_cert_public | <**ROOT**>/settings/cert_public.pem | DARE_TLS_CERT_PUBLIC | Public certificate for secure connection | security | NOT IMPLEMENTED|

**Reply by dmarro89** [reply]
Thanks!
Maybe is better to have the env name without the GO- prefix.

**Reply by vdmitriyev** [reply]
You are right. I had remove it

**Reply by dmarro89** [reply]
Initially I'd like to give the opportunity to run a http and https server. For this reason a DARE_TLS_ENABLED env variable should be added.

**Reply by vdmitriyev** [reply]
Thanks. I am going to add it to the table with configs.

**Comment by go-while** [comment]


**Reply by vdmitriyev** [reply]
This feature was not even release. Thus, it would be hard to answer the questions on viper. However, it was quite convenient to work with the package. Besides that it is quite widespread and used by many.

I try to avoid using JSON for a simple configurations, which could be managed by key/value pairs. JSON is not easy to use for configuration, because it is hard to leave comments in the file, add new/remove parameters and it could get overloaded and unmaintainable very fast.

**Reply by vdmitriyev** [reply]
Typesafe is an amazing feature when configs of one services are made for other services to consume.
For people it is usually the simpler, the better. For instance, it is easy to go with something very simple - environment variables.

**Comment by dmarro89** [comment]
@vdmitriyev  Do you think we can close the discussion or we still need it ?

**Reply by vdmitriyev** [reply]
I would keep the discussion open for a now, due to the reason, that the development process is still progressing and configs/settings could change any time (e.g., we could add a parameter to configure expiration time for a JWT token).
